0004-dockerised-local-network.md

```diff
@@ -0,0 +1,68 @@
+
+## Test plan and functional requirements
+
+1. **Testing goals and objectives**: The primary goal is to ensure that the Docker image can successfully spin up a Mina local network and that this network behaves as expected. This includes fast startup and syncing times, low resource consumption, and the ability to form a network with diverse types of participants.
+2. **Testing approach**: The testing will involve the manual functional testing only, at least for now, because of the low scripts complexity. Performance monitoring will also be conducted to ensure that the Docker image meets the resource efficiency requirements.
+3. **Testing scope**: The testing will cover all major functionalities of the Docker image, including Dockerization phase, and the resulting Mina local network operation.
+4. **Testing requirements**: The testing will require a suitable environment for running Docker, as well as tools for monitoring resource usage and blockchain behavior.
+5. **Testing resources**: The testing will be conducted on a local machine. The testing may also require additional tools for monitoring and debugging.
+
+## Drawbacks
+
+One drawback of this approach is the potential for discrepancies between the behavior of the lightweight Mina local network and the full Mina network. While the lightweight network is designed to be "close to real", it may not perfectly replicate all aspects of the full network, which could lead to misleading test results. To mitigate this drawback, we can state clearly in the documentation that the lightweight network is for local development only and is not intended to be a perfect replica of the full network. Mina local networks do not guarantee production-ready results. Users and developers must always double-check their work against real networks.
+
+Another potential drawback is the need for Docker as a dependency. Users who do not have Docker installed or who are unfamiliar with Docker may face additional hurdles in setting up and using the Mina local network.
+
+## Rationale and alternatives
+
+The Dockerization of Mina local networks is proposed as a solution to the current challenges in deploying and running local Mina networks for testing purposes. By encapsulating the Mina daemon and other necessary components in a Docker image, we can provide a simple, fast, and resource-efficient solution for spinning up local Mina networks.
+
+An alternative to this approach would be to improve the existing methods for deploying and running Mina networks, such as optimizing the setup process or providing better support for cloud-based solutions. However, these alternatives would likely be more complex and time-consuming to implement and would not offer the same level of simplicity and efficiency as the Docker-based solution.
```

Comment:
Viable Systems develop the [OpenMina](https://openmina.com/) node, which targets web / JS environments.

This goes a bit beyond this RFC, but I think it would be an interesting alternative to use OpenMina instead of the OCaml Mina node, especially for when we want to package the local Mina node as part of JS tooling like the zkapp CLI.

Reply:
@mitschabaude thank you for review and comments!

> avoids running multiple nodes altogether?
> but still produces blocks?

Yes we can have it, new implementation (or packaging) runs `one-node-network` but still behaves like complete blockchain.

> has no expensive startup phase?

Startup time dropped from 4+ minutes down to 1-1.5 minutes.

> uses less than 2GB of RAM?

Consumes `~625-650Mb` of RAM after initial startup spike.

> This goes a bit beyond this RFC, but I think it would be an interesting alternative to use OpenMina instead of the OCaml Mina node, especially for when we want to package the local Mina node as part of JS tooling like the zkapp CLI.

It will be definitely good to see what we can get from `OpenMina`.